CMSgov/design-system
# [NO-TICKET] Upgrade React types to fix Storybook 8 branch

## Summary

- Started out by upgrading the React types to the latest, trying to match `@types/react` and `@types/react-dom` as close as possible and made sure it was consistent everywhere
- Then I started trying to knock out the 109 TypeScript errors
- I found partway through that many of them could be reduced by cleaning up the `yarn.lock` file manually to get rid of old entries
- I went back and reverted a few of my changes, but many of them were actually necessary
- I changed the way the Dialog and Drawer stories are organized so that none of them pop up automatically. I then went back and added browser interaction tests to get the screenshots we want. Turns out that we were missing dialog screenshots for a while.

### The main changes

1. Upgraded React types:
    ```diff
    -    "@types/react": "^18.2.6",
    -    "@types/react-dom": "^17.0.10"
    +    "@types/react": "^18.3.1",
    +    "@types/react-dom": "^18.3.0"
    ```
2. Explicitly told Storybook to start using `react-docgen-typescript`. If my fix for union-member sort order ever gets released, we'll need to follow up and change this config to pass the new option to that library.

## How to test

1. `yarn type-check`
3. `yarn storybook`
4. Anything else you can think to check

## Checklist

Planning to merge it into the other branch, and I think we'll want to put that branch into v11
